File: packages/soulseek_protocol/lib/src/transfer/download_manager.dart
import 'dart:async';
import 'dart:io';

import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';
import '../peer/peer_connection.dart';

enum DownloadState { queued, requesting, downloading, paused, completed, failed, cancelled }

class DownloadFile {
  final String filename;
  final int size;
  final String username;
  final int fileCode;
  final String? localPath;

  DownloadState state;
  int downloadedBytes;
  int position;
  String? error;
  double? speed;
  String? placeInQueue;
  int retryCount = 0;

  final _progressController = StreamController<DownloadProgress>.broadcast();

  DownloadFile({
    required this.filename,
    required this.size,
    required this.username,
    required this.fileCode,
    this.localPath,
    this.state = DownloadState.queued,
    this.downloadedBytes = 0,
    this.position = 0,
  });

  Stream<DownloadProgress> get progress => _progressController.stream;

  void emitProgress() {
    _progressController.add(DownloadProgress(
      filename: filename,
      downloadedBytes: downloadedBytes,
      totalSize: size,
      speed: speed,
      state: state,
    ));
  }

  void dispose() {
    _progressController.close();
  }
}

class DownloadProgress {
  final String filename;
  final int downloadedBytes;
  final int totalSize;
  final double? speed;
  final DownloadState state;

  const DownloadProgress({
    required this.filename,
    required this.downloadedBytes,
    required this.totalSize,
    this.speed,
    required this.state,
  });

  double get percentage => totalSize > 0 ? downloadedBytes / totalSize : 0;
}

class DownloadManager {
  final Map<String, DownloadFile> _downloads = {};
  final _progressController = StreamController<Map<String, DownloadProgress>>.broadcast();
  Timer? _speedTimer;
  int _activeCount = 0;

  final int maxConcurrent;
  final int maxRetries;
  final Map<String, PeerConnection> _activeConnections = {};

  DownloadManager({
    this.maxConcurrent = 3,
    this.maxRetries = 3,
  });

  Stream<Map<String, DownloadProgress>> get allProgress => _progressController.stream;
  int get activeDownloadCount => _activeCount;
  int get queuedDownloadCount => _downloads.values.where((d) => d.state == DownloadState.queued).length;

  DownloadFile addDownload({
    required String filename,
    required int size,
    required String username,
    required int fileCode,
    String? localPath,
  }) {
    final download = DownloadFile(
      filename: filename,
      size: size,
      username: username,
      fileCode: fileCode,
      localPath: localPath ?? _defaultPath(filename),
    );
    _downloads['$username:$filename'] = download;
    _emitProgress();
    _processQueue();
    return download;
  }

  Future<void> startDownload(DownloadFile download, PeerConnection connection) async {
    if (download.state != DownloadState.queued && download.state != DownloadState.failed) return;

    _activeCount++;
    download.state = DownloadState.requesting;
    _activeConnections['${download.username}:${download.filename}'] = connection;
    _emitProgress();

    try {
      final request = TransferRequest(
        direction: Direction.download,
        fileCode: download.fileCode,
        filename: download.filename,
        fileSize: download.size,
      );
      connection.sendMessage(SoulseekMessage(request.code, request.serialize().toBytes()));

      download.state = DownloadState.downloading;
      _emitProgress();

      await _receiveFile(download, connection);
    } catch (e) {
      download.state = DownloadState.failed;
      download.error = e.toString();
      if (download.retryCount < maxRetries) {
        download.retryCount++;
        download.state = DownloadState.queued;
      }
      _emitProgress();
    } finally {
      _activeCount--;
      _activeConnections.remove('${download.username}:${download.filename}');
      _processQueue();
    }
  }

  Future<void> _receiveFile(DownloadFile download, PeerConnection connection) async {
    final file = File(download.localPath!);
    final raf = await file.open(mode: FileMode.write);
    int lastBytes = 0;
    final stopwatch = Stopwatch()..start();

    try {
      final sub = connection.messages.listen((message) async {
        if (message.code == PeerCode.transferResponse) {
          final buffer = ReadBuffer(message.payload);
          final response = buffer.readInt32();
          if (response != 0) {
            download.state = DownloadState.failed;
            download.error = 'Transfer denied (code: $response)';
            download.emitProgress();
            return;
          }
        }

        raf.writeFromSync(message.payload);
        download.downloadedBytes += message.payload.length;
        download.position += message.payload.length;

        if (stopwatch.elapsedMilliseconds >= 500) {
          final elapsed = stopwatch.elapsedMilliseconds / 1000.0;
          download.speed = (download.downloadedBytes - lastBytes) / elapsed;
          lastBytes = download.downloadedBytes;
          stopwatch.reset();
        }

        _emitProgress();
        download.emitProgress();

        if (download.downloadedBytes >= download.size) {
          download.state = DownloadState.completed;
          _emitProgress();
          download.emitProgress();
        }
      });

      await Future.doWhile(() async {
        await Future.delayed(const Duration(milliseconds: 100));
        return download.state == DownloadState.downloading;
      });

      await sub.cancel();
    } finally {
      await raf.close();
    }
  }

  void pauseDownload(DownloadFile download) {
    if (download.state == DownloadState.downloading) {
      download.state = DownloadState.paused;
      _emitProgress();
      download.emitProgress();
    }
  }

  void resumeDownload(DownloadFile download) {
    if (download.state == DownloadState.paused) {
      download.state = DownloadState.queued;
      _emitProgress();
      _processQueue();
    }
  }

  void cancelDownload(DownloadFile download) {
    download.state = DownloadState.cancelled;
    _emitProgress();
    download.emitProgress();
  }

  void removeDownload(DownloadFile download) {
    _downloads.remove('${download.username}:${download.filename}');
    _activeConnections.remove('${download.username}:${download.filename}');
    download.dispose();
    _emitProgress();
  }

  void retryDownload(DownloadFile download) {
    if (download.state == DownloadState.failed) {
      download.retryCount = 0;
      download.state = DownloadState.queued;
      _emitProgress();
      _processQueue();
    }
  }

  void _processQueue() {
    if (_activeCount >= maxConcurrent) return;

    final queued = _downloads.values
        .where((d) => d.state == DownloadState.queued)
        .toList();

    for (final download in queued) {
      if (_activeCount >= maxConcurrent) break;
      final key = '${download.username}:${download.filename}';
      final conn = _activeConnections[key];
      if (conn != null) {
        _activeCount++;
        download.state = DownloadState.requesting;
        _emitProgress();
        _startQueuedDownload(download, conn);
      }
    }
  }

  Future<void> _startQueuedDownload(DownloadFile download, PeerConnection connection) async {
    try {
      await startDownload(download, connection);
    } catch (e) {
      download.state = DownloadState.failed;
      download.error = e.toString();
      _emitProgress();
    }
  }

  void _emitProgress() {
    final progressMap = <String, DownloadProgress>{};
    for (final entry in _downloads.entries) {
      final d = entry.value;
      progressMap[entry.key] = DownloadProgress(
        filename: d.filename,
        downloadedBytes: d.downloadedBytes,
        totalSize: d.size,
        speed: d.speed,
        state: d.state,
      );
    }
    _progressController.add(progressMap);
  }

  String _defaultPath(String filename) {
    final name = filename.split(Platform.pathSeparator).last;
    final dir = Directory.systemTemp.path;
    return '$dir/$name';
  }

  void dispose() {
    _speedTimer?.cancel();
    _progressController.close();
    for (final d in _downloads.values) {
      d.dispose();
    }
    for (final conn in _activeConnections.values) {
      conn.disconnect();
    }
    _downloads.clear();
    _activeConnections.clear();
  }
}

File: Soulseek.Protocol/Transfer/DownloadManager.cs
using Soulseek.Protocol.Messages;
using Soulseek.Protocol.Peer;

namespace Soulseek.Protocol.Transfer;

public enum DownloadState
{
    Queued,
    Requesting,
    Downloading,
    Paused,
    Completed,
    Failed,
    Cancelled
}

public class DownloadFile
{
    public string Filename { get; }
    public int Size { get; }
    public string Username { get; }
    public int FileCode { get; }
    public string? LocalPath { get; set; }

    public DownloadState State { get; set; }
    public int DownloadedBytes { get; set; }
    public int Position { get; set; }
    public string? Error { get; set; }
    public double? Speed { get; set; }
    public string? PlaceInQueue { get; set; }
    public int RetryCount { get; set; }

    private readonly Subject<DownloadProgress> _progressSubject = new();

    public DownloadFile(string filename, int size, string username, int fileCode, string? localPath = null,
        DownloadState state = DownloadState.Queued)
    {
        Filename = filename;
        Size = size;
        Username = username;
        FileCode = fileCode;
        LocalPath = localPath ?? DefaultPath(filename);
        State = state;
    }

    public IObservable<DownloadProgress> Progress => _progressSubject;

    public void EmitProgress()
    {
        _progressSubject.OnNext(new DownloadProgress(Filename, DownloadedBytes, Size, Speed, State));
    }

    private static string DefaultPath(string filename)
    {
        var name = Path.GetFileName(filename) ?? filename;
        return Path.Combine(Path.GetTempPath(), name);
    }

    public void Dispose()
    {
        _progressSubject.Dispose();
    }
}

public record DownloadProgress(
    string Filename,
    int DownloadedBytes,
    int TotalSize,
    double? Speed,
    DownloadState State
)
{
    public double Percentage => TotalSize > 0 ? (double)DownloadedBytes / TotalSize : 0;
}

public class DownloadManager
{
    private readonly Dictionary<string, DownloadFile> _downloads = new();
    private readonly Subject<Dictionary<string, DownloadProgress>> _progressSubject = new();
    private readonly Dictionary<string, PeerConnection> _activeConnections = new();

    public int MaxConcurrent { get; }
    public int MaxRetries { get; }
    private int _activeCount;

    public DownloadManager(int maxConcurrent = 3, int maxRetries = 3)
    {
        MaxConcurrent = maxConcurrent;
        MaxRetries = maxRetries;
    }

    public IObservable<Dictionary<string, DownloadProgress>> AllProgress => _progressSubject;
    public int ActiveDownloadCount => _activeCount;
    public int QueuedDownloadCount => _downloads.Values.Count(d => d.State == DownloadState.Queued);

    public DownloadFile AddDownload(string filename, int size, string username, int fileCode, string? localPath = null)
    {
        var download = new DownloadFile(filename, size, username, fileCode, localPath);
        _downloads[$"{username}:{filename}"] = download;
        EmitProgress();
        ProcessQueue();
        return download;
    }

    public async Task StartDownload(DownloadFile download, PeerConnection connection)
    {
        if (download.State != DownloadState.Queued && download.State != DownloadState.Failed) return;

        _activeCount++;
        download.State = DownloadState.Requesting;
        _activeConnections[$"{download.Username}:{download.Filename}"] = connection;
        EmitProgress();

        try
        {
            var request = new TransferRequest(Direction.Download, download.FileCode, download.Filename, download.Size);
            connection.SendMessage(new SoulseekMessage(request.Code, request.Serialize().ToBytes()));

            download.State = DownloadState.Downloading;
            EmitProgress();

            await ReceiveFile(download, connection);
        }
        catch (Exception e)
        {
            download.State = DownloadState.Failed;
            download.Error = e.Message;
            if (download.RetryCount < MaxRetries)
            {
                download.RetryCount++;
                download.State = DownloadState.Queued;
            }
            EmitProgress();
        }
        finally
        {
            _activeCount--;
            _activeConnections.Remove($"{download.Username}:{download.Filename}");
            ProcessQueue();
        }
    }

    private async Task ReceiveFile(DownloadFile download, PeerConnection connection)
    {
        var filePath = download.LocalPath ?? Path.Combine(Path.GetTempPath(), Path.GetFileName(download.Filename));
        await using var fileStream = new FileStream(filePath, FileMode.Create, FileAccess.Write);
        int lastBytes = 0;
        var stopwatch = System.Diagnostics.Stopwatch.StartNew();
        var responseReceived = false;

        var tcs = new TaskCompletionSource();

        using var sub = connection.Messages.Subscribe(message =>
        {
            if (message.Code != PeerCode.TransferResponse) return;

            try
            {
                if (!responseReceived)
                {
                    responseReceived = true;
                    if (message.Payload.Length >= 4)
                    {
                        var response = BitConverter.ToInt32(message.Payload, 0);
                        if (response != 0)
                        {
                            download.State = DownloadState.Failed;
                            download.Error = $"Transfer denied (code: {response})";
                            download.EmitProgress();
                            EmitProgress();
                            tcs.TrySetResult();
                            return;
                        }
                    }
                    return;
                }

                fileStream.Write(message.Payload, 0, message.Payload.Length);
                download.DownloadedBytes += message.Payload.Length;
                download.Position += message.Payload.Length;

                if (stopwatch.ElapsedMilliseconds >= 500)
                {
                    var elapsed = stopwatch.ElapsedMilliseconds / 1000.0;
                    download.Speed = (download.DownloadedBytes - lastBytes) / elapsed;
                    lastBytes = download.DownloadedBytes;
                    stopwatch.Restart();
                }

                EmitProgress();
                download.EmitProgress();

                if (download.DownloadedBytes >= download.Size)
                {
                    download.State = DownloadState.Completed;
                    EmitProgress();
                    download.EmitProgress();
                    tcs.TrySetResult();
                }
            }
            catch (Exception ex)
            {
                if (!tcs.Task.IsCompleted)
                {
                    download.State = DownloadState.Failed;
                    download.Error = ex.Message;
                    tcs.TrySetResult();
                }
            }
        });

        await tcs.Task;
    }

    public void PauseDownload(DownloadFile download)
    {
        if (download.State == DownloadState.Downloading)
        {
            download.State = DownloadState.Paused;
            EmitProgress();
            download.EmitProgress();
        }
    }

    public void ResumeDownload(DownloadFile download)
    {
        if (download.State == DownloadState.Paused)
        {
            download.State = DownloadState.Queued;
            EmitProgress();
            ProcessQueue();
        }
    }

    public void CancelDownload(DownloadFile download)
    {
        download.State = DownloadState.Cancelled;
        EmitProgress();
        download.EmitProgress();
    }

    public void RemoveDownload(DownloadFile download)
    {
        _downloads.Remove($"{download.Username}:{download.Filename}");
        _activeConnections.Remove($"{download.Username}:{download.Filename}");
        download.Dispose();
        EmitProgress();
    }

    public void RetryDownload(DownloadFile download)
    {
        if (download.State == DownloadState.Failed)
        {
            download.RetryCount = 0;
            download.State = DownloadState.Queued;
            EmitProgress();
            ProcessQueue();
        }
    }

    private void ProcessQueue()
    {
        if (_activeCount >= MaxConcurrent) return;

        var queued = _downloads.Values.Where(d => d.State == DownloadState.Queued).ToList();

        foreach (var download in queued)
        {
            if (_activeCount >= MaxConcurrent) break;
            var key = $"{download.Username}:{download.Filename}";
            if (_activeConnections.TryGetValue(key, out var conn))
            {
                _activeCount++;
                download.State = DownloadState.Requesting;
                EmitProgress();
                _ = StartQueuedDownload(download, conn);
            }
        }
    }

    private async Task StartQueuedDownload(DownloadFile download, PeerConnection connection)
    {
        try
        {
            await StartDownload(download, connection);
        }
        catch (Exception e)
        {
            download.State = DownloadState.Failed;
            download.Error = e.Message;
            EmitProgress();
        }
    }

    private void EmitProgress()
    {
        var progressMap = new Dictionary<string, DownloadProgress>();
        foreach (var entry in _downloads)
        {
            var d = entry.Value;
            progressMap[entry.Key] = new DownloadProgress(d.Filename, d.DownloadedBytes, d.Size, d.Speed, d.State);
        }
        _progressSubject.OnNext(progressMap);
    }

    public void Dispose()
    {
        _progressSubject.Dispose();
        foreach (var d in _downloads.Values)
            d.Dispose();
        _downloads.Clear();
        _activeConnections.Clear();
    }
}